src/components/menuFilter/menu-filter.js

```diff
@@ -0,0 +1,146 @@
+/**
+ * @ngdoc module
+ * @name material.components.menuFilter
+ */
+
+angular.module('material.components.menuFilter', [
+  'material.core',
+])
+  .directive('mdMenuFilter', MdMenuFilterDirective);
+
+/**
+ * @ngdoc directive
+ * @name mdMenuFilter
+ * @module material.components.menuFilter
+ * @restrict E
+ *
+ * @description
+ * The menu filter directive is used to make adding a searchbar to
+ * the top of a component quick and painless.
+ *
+ * @param {string=} aria-label Optional label for accessibility. Only necessary if no placeholder or
+ *   explicit label is present.
+ * @param {Array=} filtered-items The variable the filtered
+ *   results will be stored in.
+ * @param {Array=} items The items to filter.
+ * @param {string} ng-model Assignable angular expression to data-bind the
+ *   search term to.
+ * @param {string=} placeholder Placeholder hint text.
+ * @param {string=} tabindex Optional tab-index for accessibility.
+ *
+ * @usage
+ * <hljs lang="html">
+ *  <md-menu-filter items="ctrl.items"
+ *                  placeholder="Select an item.."
+ *                  filtered-items="ctrl.filteredItems"
+ *                  ng-model="ctrl.searchTerm"
+ *                  tabindex="1"
+ *                  aria-label="search box">
+ *  </md-menu-filter>
+ * </hljs>
+ */
+function MdMenuFilterDirective($mdTheming, $mdAria) {
+  return {
+    restrict: 'E',
+    template: '<input ng-model="searchTerm" type="search"' +
+        'class="md-menu-filter _md-text">' +
+            '<button' +
+                ' type="button"' +
+                ' tabindex="-1"' +
+                ' ng-if="searchTerm"' +
+                ' class="search-clear-button"' +
+                ' ng-click="clear()">' +
+              '<md-icon md-svg-icon="md-close"></md-icon>' +
+              '<span class="_md-visually-hidden">Clear</span>' +
+            '</button>',
+    scope: {
+      filteredItems: '=',
+      items: '=',
+      ngModel: '=',
+    },
+    link: link
+  };
+
+  function link(scope, element, attr) {
+    var ngModel = element.controller('ngModel');
+
+    $mdTheming(element);
+
+    init();
+
+    function init() {
+      // This is for the clear button that appears when there is a search term.
+      scope.clear = function() {
+        scope.searchTerm = '';
+      };
+
+      // Set our tabindex of the md-menu-filter directive to -1 by default,
+      // because the select/menu input will hold the actual tabindex.
+      element.find('input').attr('tabindex', attr.tabindex || '-1');
```

Comment:
Hmmm, this means that this component _must_ be used within a menu or select (really only select since menus shouldn't allow search). I had thought that we wanted to make something generic that I could even use it within my custom component outside of material.